File: Assets/Scripts/Core/Offline.cs
// Offline.cs — Offline-progressie-engine (vertaling van src/core/offline.ts).
// Puur, geen MonoBehaviour, immutability via Clone(). Bouwt op BankHelper, Progression, DropSystem, Bonuses.

using System;
using System.Collections.Generic;
using IdleCloud.Data;

namespace IdleCloud.Core
{
    public static class Offline
    {
        // ── Constanten (spiegelen TS-bron exact) ─────────────────────────────

        // ── Publieke API ──────────────────────────────────────────────────────

        /// <summary>
        /// Runt de offline-formule voor alle personages en stort loot/coins in de gedeelde bank.
        /// Puur: geeft een nieuw Account + een OfflineReport terug (null als er te weinig tijd is
        /// verstreken of geen enkel personage iets deed).
        /// Spiegelt simulateOffline() uit offline.ts exact.
        /// </summary>
        public static (Account Account, OfflineReport Report) SimulateOffline(
            Account account,
            long now,
            OfflineDataBundle data,
            IRandomSource rng,
            AccountBonuses bonuses = null)
        {
            if (account == null) throw new ArgumentNullException(nameof(account));
            if (data == null) throw new ArgumentNullException(nameof(data));
            if (rng == null) throw new ArgumentNullException(nameof(rng));
            if (account.Bank == null) throw new ArgumentException("Account bank is required.", nameof(account));
            if (bonuses == null) bonuses = AccountBonuses.Zero();
            if (now < account.LastSeenAt) throw new ArgumentOutOfRangeException(nameof(now), "Time cannot move backwards.");

            OfflineBalanceConfig balance = data.OfflineBalance ?? OfflineBalanceRepo.Default;
            if (balance == null || balance.Rate < 0 || balance.Rate > 1 ||
                balance.CapMs <= 0 || balance.MinimumDurationMs < 0 ||
                balance.MinimumDurationMs > balance.CapMs)
                throw new ArgumentException("Offline balance is invalid.", nameof(data));

            long elapsedMs = now - account.LastSeenAt;
            if (elapsedMs < balance.MinimumDurationMs)
            {
                var quickClone      = CloneAccount(account);
                quickClone.LastSeenAt = now;
                return (quickClone, null);
            }

            long cappedMs = System.Math.Min(elapsedMs, balance.CapMs);
            double hours  = cappedMs / 3_600_000.0;

            Bank bank                          = account.Bank.Clone();
            var characters                     = new List<Character>();
            var reports                        = new List<OfflineCharacterReport>();
            double dropMult                    = 1.0 + bonuses.DropPct;

            foreach (var c in account.Characters ?? new List<Character>())
            {
                var (simChar, report) = SimulateCharacter(c, hours, data, rng, dropMult, balance.Rate);

                // Loot gaat eerst naar de inventaris van het personage
                var lootOverflow = new List<ItemStack>();
                var updatedChar = simChar.Clone();
                foreach (var stack in report.Loot)
                {
                    if (!data.Items.TryGetValue(stack.ItemId, out ItemDef def)) continue;
                    var (newCharacter, overflow) = Inventory.AddToInventory(updatedChar, stack, def);
                    updatedChar = newCharacter;
                    if (overflow > 0)
                        lootOverflow.Add(new ItemStack { ItemId = stack.ItemId, Qty = overflow });
                }
                if (report.CoinsGained > 0)
                    bank = BankHelper.AddCoins(bank, report.CoinsGained);

                // Herstel startedAt zodat de volgende heartbeat correct werkt
                updatedChar.Activity   = new ActivityState
                {
                    Kind      = simChar.Activity?.Kind ?? ActivityKind.Idle,
                    TargetId  = simChar.Activity?.TargetId,
                    StartedAt = now,
                };
                characters.Add(updatedChar);

                report.LootOverflow = lootOverflow;
                reports.Add(report);
            }

            bool didAnything = false;
            foreach (var r in reports)
                if (r.Actions > 0) { didAnything = true; break; }

            var resultAccount            = CloneAccount(account);
            resultAccount.Characters     = characters;
            resultAccount.Bank           = bank;
            resultAccount.LastSeenAt     = now;

            OfflineReport offlineReport  = didAnything
                ? new OfflineReport { ElapsedMs = elapsedMs, CappedMs = cappedMs, Characters = reports }
                : null;

            return (resultAccount, offlineReport);
        }

        // ── Private helpers ───────────────────────────────────────────────────

        /// <summary>
        /// Verwerkt één personage voor de offline-periode en retourneert het bijgewerkte personage
        /// + het per-personage-rapport. Spiegelt simulateCharacter() uit offline.ts exact.
        /// </summary>
        private static (Character Character, OfflineCharacterReport Report) SimulateCharacter(
            Character character,
            double hours,
            OfflineDataBundle data,
            IRandomSource rng,
            double dropMult,
            double offlineRate)
        {
            var emptyReport = new OfflineCharacterReport
            {
                CharacterId   = character.Id,
                CharacterName = character.Name,
                Kind          = character.Activity?.Kind ?? ActivityKind.Idle,
                TargetId      = character.Activity?.TargetId,
                Actions       = 0,
                XpGained      = 0,
                LevelsGained  = 0,
                CoinsGained   = 0,
                Loot          = new List<ItemStack>(),
                LootOverflow  = new List<ItemStack>(),
            };

            EfficiencySnapshot snap = character.Efficiency;
            if (!SnapshotValidation.IsUsable(snap, character) || snap.Kind == ActivityKind.Idle || snap.ActionsPerHour <= 0)
                return (character, emptyReport);

            double rawActions = snap.ActionsPerHour * offlineRate * hours;
            int actions = rawActions >= int.MaxValue
                ? int.MaxValue
                : (int)System.Math.Floor(rawActions);
            if (actions <= 0) return (character, emptyReport);

            long xpGained     = (long)System.Math.Floor(actions * snap.XpPerAction);
            int  coinsGained  = (int)System.Math.Floor(actions * snap.CoinsPerAction);

            // Combat → gewogen OSRS-tabel; life-skill nodes → kans-gebaseerd model
            List<ItemStack> loot;
            if (snap.Kind == ActivityKind.Fighting)
            {
                DropTable table = null;
                if (snap.TargetId != null && data.Monsters.TryGetValue(snap.TargetId, out MonsterDef mon))
                    table = mon.Drops;
                loot = table != null
                    ? DropSystem.ExpectedDropTable(table, actions, rng, dropMult - 1.0)
                    : new List<ItemStack>();
            }
            else
            {
                List<DropEntry> nodeDrops = new List<DropEntry>();
                if (snap.TargetId != null && data.Nodes.TryGetValue(snap.TargetId, out ResourceNodeDef node))
                    nodeDrops = node.Drops ?? new List<DropEntry>();
                loot = DropSystem.RollDrops(nodeDrops, actions, rng, dropMult);
            }

            Character next = character.Clone();
            int levelsGained;

            if (snap.Kind == ActivityKind.Fighting)
            {
                // Gevechts-XP gaat naar zowel het personage-level als de combat-vaardigheid
                var charResult = CharacterHelper.ApplyCharacterXp(next, xpGained);
                next = charResult.Character;
                levelsGained = charResult.LevelsGained;
                next = CharacterHelper.ApplySkillXp(next, SkillId.Combat, xpGained).Character;
            }
            else
            {
                SkillId skillId = ActivitySkillMapping.ToSkillId(snap.Kind);
                var skillResult = CharacterHelper.ApplySkillXp(next, skillId, xpGained);
                next = skillResult.Character;
                levelsGained = skillResult.LevelsGained;
            }

            var report = new OfflineCharacterReport
            {
                CharacterId   = emptyReport.CharacterId,
                CharacterName = emptyReport.CharacterName,
                Kind          = emptyReport.Kind,
                TargetId      = emptyReport.TargetId,
                Actions       = actions,
                XpGained      = xpGained,
                LevelsGained  = levelsGained,
                CoinsGained   = coinsGained,
                Loot          = loot,
                LootOverflow  = new List<ItemStack>(), // wordt ingevuld door simulateOffline
            };

            return (next, report);
        }

        /// <summary>Ondiepe kopie van Account voor immutable updates.</summary>
        private static Account CloneAccount(Account account) => account.Clone();
    }
}

File: webapp/src/domain/simulation/offline.ts
import type {
  Account,
  ContentBundle,
  ItemStack,
  OfflineBalanceConfig,
  OfflineCharacterReport,
  OfflineReport
} from "../types";
import { expectedDropTable, rollDrops } from "./dropSystem";
import { applyCharacterXp } from "./progression";

function hashSeed(parts: Array<string | number>): number {
  const raw = parts.join("|");
  let hash = 2166136261;
  for (let i = 0; i < raw.length; i += 1) {
    hash ^= raw.charCodeAt(i);
    hash = Math.imul(hash, 16777619);
  }
  return hash >>> 0;
}

function createRng(seed: number): () => number {
  let state = seed || 1;
  return () => {
    state = (Math.imul(1664525, state) + 1013904223) >>> 0;
    return state / 4294967296;
  };
}

function pushStack(target: ItemStack[], itemId: string, qty: number): void {
  if (qty <= 0) return;
  const existing = target.find((entry) => entry.itemId === itemId);
  if (existing) {
    existing.qty += qty;
    return;
  }
  target.push({ itemId, qty });
}

export function simulateOffline(
  account: Account,
  nowMs: number,
  content: ContentBundle,
  balance: OfflineBalanceConfig
): { account: Account; report: OfflineReport | null } {
  if (nowMs < account.lastSeenAt) {
    throw new Error("Time moved backwards");
  }

  const elapsedMs = nowMs - account.lastSeenAt;
  if (elapsedMs < balance.minimumDurationMs) {
    return {
      account: { ...account, lastSeenAt: nowMs },
      report: null
    };
  }

  const cappedMs = Math.min(elapsedMs, balance.capMs);
  const hours = cappedMs / 3_600_000;
  const rng = createRng(hashSeed([account.id, account.lastSeenAt, nowMs]));

  const reports: OfflineCharacterReport[] = [];
  const nextCharacters = account.characters.map((character) => {
    const efficiency = character.efficiency;
    const baseReport: OfflineCharacterReport = {
      characterId: character.id,
      characterName: character.name,
      kind: character.activity.kind,
      targetId: character.activity.targetId,
      actions: 0,
      xpGained: 0,
      levelsGained: 0,
      coinsGained: 0,
      loot: []
    };

    if (!efficiency || efficiency.kind === "Idle" || efficiency.actionsPerHour <= 0) {
      reports.push(baseReport);
      return {
        ...character,
        activity: { ...character.activity, startedAt: nowMs }
      };
    }

    const actions = Math.floor(efficiency.actionsPerHour * balance.rate * hours);
    if (actions <= 0) {
      reports.push(baseReport);
      return {
        ...character,
        activity: { ...character.activity, startedAt: nowMs }
      };
    }

    const xpGained = Math.floor(actions * efficiency.xpPerAction);
    const coinsGained = Math.floor(actions * efficiency.coinsPerAction);
    let loot: ItemStack[] = [];

    if (efficiency.kind === "Fighting" && efficiency.targetId) {
      const monster = content.monsters[efficiency.targetId];
      if (monster?.dropTable) {
        loot = expectedDropTable(monster.dropTable, actions, rng, 0);
      } else {
        loot = rollDrops(monster?.drops ?? [], actions, rng, 1);
      }
    } else if (efficiency.targetId) {
      const node = content.nodes[efficiency.targetId];
      loot = rollDrops(node?.drops ?? [], actions, rng, 1);
    }

    const progression = applyCharacterXp(character, xpGained);
    const nextCharacter = {
      ...progression.character,
      inventory: [...character.inventory],
      activity: { ...character.activity, startedAt: nowMs }
    };

    for (const stack of loot) {
      pushStack(nextCharacter.inventory, stack.itemId, stack.qty);
    }

    reports.push({
      ...baseReport,
      actions,
      xpGained,
      levelsGained: progression.levelsGained,
      coinsGained,
      loot
    });

    return nextCharacter;
  });

  const totalCoins = reports.reduce((sum, entry) => sum + entry.coinsGained, 0);
  const didAnything = reports.some((entry) => entry.actions > 0);

  const nextAccount: Account = {
    ...account,
    lastSeenAt: nowMs,
    characters: nextCharacters,
    bank: {
      ...account.bank,
      coins: account.bank.coins + totalCoins
    }
  };

  if (!didAnything) {
    return { account: nextAccount, report: null };
  }

  return {
    account: nextAccount,
    report: {
      elapsedMs,
      cappedMs,
      characters: reports
    }
  };
}
